tests/celery_test_workers.py

```diff
@@ -52,6 +56,31 @@ def __init__(self, app: Celery):
         self.worker_processes = {}
         self.echo_processes = {}
 
+    def __enter__(self):
+        """This magic method is necessary for allowing this class to be used as a context manager."""
+        return self
+
+    def __exit__(self, exc_type: Type[Exception], exc_value: Exception, traceback: TracebackType):
+        """
+        This will always run at the end of a context with statement, even if an error is raised.
+        It's a safe way to ensure all of our subprocesses are stopped no matter what.
+        """
+
+        # Try to stop everything gracefully first
+        self.stop_all_workers()
+
+        # Check that all the worker processes were stopped, otherwise forcefully terminate them
+        for worker_process in self.worker_processes.values():
+            if worker_process.is_alive():
+                worker_process.kill()
+
+        # Check that all the echo processes were stopped, otherwise forcefully terminate them
+        ps_proc = subprocess.run("ps ux", shell=True, capture_output=True, text=True)
```

Comment:
will this grab just the needed processes or could it grab others too? Does this need a filter or grep?

Reply:
That will give me all the currently running processes, then on the next line, where it says echo_processes.values(), it will filter for just the processes I'm looking to terminate

Reply:
Ideally, the processes will have already been stopped by the call to stop_all_workers. This is just one additional check to make sure they're all shutdown.